spark/src/main/scala/org/apache/spark/sql/delta/DeltaAnalysis.scala

```diff
@@ -917,42 +923,46 @@ class DeltaAnalysis(session: SparkSession)
       attr: NamedExpression,
       targetAttr: NamedExpression,
       tblName: String,
-      allowTypeWidening: Boolean): NamedExpression = {
+      shouldWidenType: (AtomicType, AtomicType) => Boolean): NamedExpression = {
```

Comment:
I'm slightly concerned about this signature for the same reason I mentioned above: Both types are the same and it's easy to mess them up. Ofc from -> to is logical for you an me. But that doesn't mean it'll be for everyone. It would be good to be explicit to harden this code. So I just making this a single method trait instead that can be implemented differently at various places. But then at least the method has the correct names and can be invoked with by-name-args.

Reply:
I introduced `TypeWideningMode` following your suggestion